picard/tagger.py

```diff
@@ -186,6 +187,9 @@ def __init__(self, items):
             parsed = urlparse(item)
             if not parsed.scheme:
                 self.files.add(item)
+            elif parsed.scheme == "command":
+                for x in item.replace("command://", '').upper().split(';'):
```

Comment:
I wonder about how this would play if we want to pass arguments along commands: "CMD1 ARG1 ARG2; CMD2 ARG3"
It might be useful, for example to target a specific cluster or album.
Also I'm not sure about the handling of multiple commands using ";" separator, instead of repeating "-e" option.
@phw : what do you think about allowing arguments?


I don't think we should convert to uppercase here, but instead do case-insensitive comparison in `handle_command()`.

`item.replace("command://", '')` -> `item[10:]`

Reply:
I agree that repeating the -e option with command and argument to use makes more sense from a user perspective.

Reply:
I agree with the repeating -e, good point.

I’m not sure what do you mean by a *case-insensitive comaprison* tho @zas

Reply:
I meant just do the conversion when needed.

```python
    def handle_command(self, command):
        command = command.upper()
        if command == "SHOW":
            self.bring_tagger_front()
        elif command == "QUIT":
            self.exit()
```

BTW, I was thinking about something like:

```python
    self.commands = {
        'SHOW': self.handle_command_show,
        'QUIT': self.handle_command_quit,
        'CMD_WITH_ARGS': self.handle_command_with_args,
    }

...

    def handle_command(self, command):
        try:
            cmd, argstring = command.strip().split(" ", 1)
            commands[cmd.upper()](argstring.strip())
        except KeyError:
            log.error("Unknown command: %r", command)

    def handle_command_show(self, argstring):
        self.bring_tagger_front()

    def handle_command_quit(self, argstring):
        self.exit()

    def handle_command_with_args(self, argstring):
        if not argstring:
            log.error("Missing argument")
        do_something_with(argstring)
```

(indentation may be incorrect, just an example)

It would be easier to extend, and commands can be added dynamically.

Reply:
When splitting, you may also need to trap `ValueError: not enough values to unpack (expected 2, got 1)` in case there is no space (no args for the command).

Reply:
Okay, there's a problem with your idea:

```
〉picard -e a -e b -e c
c
```

Basically it's a flag that can have only one value. Filepaths/URLs are positional arguments, so I don't see the other way than `picard -e "SHOW; QUIT"`

Reply:
The problem is that we cannot use a flag (which is `-e/--exec`) more than once because it gets overriden. So even if we add `nargs='*'`, we end up with something like this:

```
〉picard -e a -e b -e c
['c']

〉picard -e a b c
['a', 'b', 'c']
```

So we've returned to the start point, in that case I stick to my `picard -e "SHOW; CLUSTER; QUIT"` idea

Reply:
> The problem is that we cannot use a flag (which is -e/--exec) more than once because it gets overriden.

Try this:

```python
>>> p = argparse.ArgumentParser()
>>> p.add_argument("-e", nargs='+', action='append')
_AppendAction(option_strings=['-e'], dest='e', nargs='+', const=None, default=None, type=None, choices=None, help=None, metavar=None)
>>> p.parse_args("-e a b c -e d e -e f".split())
Namespace(e=[['a', 'b', 'c'], ['d', 'e'], ['f']])
>>> p.parse_args("-e".split())
usage: [-h] [-e E [E ...]]
: error: argument -e: expected at least one argument
```

Reply:
Aight, thanks